Allow for pooling of empty batch (like for conv in #3715).

lucasb-eyer opened the pull request:
Similarly to #3715 allowing size-zero batches to be sent through cuDNN conv (essentially bypassing the call to cuDNN and allocating a size-zero batch output of the correct size), this allows size-zero batches to be sent through cuDNN pooling.

Comes with unit-tests and I have used this change in a real-world application for a couple weeks already.

nouiz commented:
jenkins test this.

Good timing, we want to do an RC tomorrow.

nouiz commented on theano/sandbox/cuda/dnn.py:
```diff
@@ -1710,6 +1707,24 @@ def c_code(self, node, name, inputs, outputs, sub):
+// (which will fail on zero batch size)
+if (CudaNdarray_DIMS(%(input)s)[0] == 0) {
+  cudaError_t err2 = cudaMemset((%(out)s)->devdata, 0,
+                                CudaNdarray_SIZE(%(out)s) * sizeof(real));
```
This call is useless, as out will have a shape of (0, ....) and so the total array size is 0.

nouiz commented on theano/sandbox/cuda/dnn.py:
```diff
@@ -1731,6 +1746,8 @@ def c_code(self, node, name, inputs, outputs, sub):
   %(fail)s
 }
+
+}
```
As you didn't changed the indentation, add a comment to tell which if it is matching to help reader of the code.

nouiz commented on theano/sandbox/cuda/dnn.py:
```diff
@@ -1937,20 +1954,34 @@ def c_code(self, node, name, inputs, outputs, sub):
+// (which will fail on zero batch size)
+if (CudaNdarray_DIMS(%(input)s)[0] == 0) {
+  cudaError_t err2 = cudaMemset((%(output)s)->devdata, 0,
+                                CudaNdarray_SIZE(%(output)s) * sizeof(real));
```
idem here and for the comment.

lucasb-eyer commented:
Haha no worries I didn't realize about the RC! No need to squeeze this in.

Good point about not needing to memset size-0 arrays. This was taken over from the conv-code linked in #3751, so I guess it's superflous there too. I changed the control flow accordingly; tests still pass on my machine. Also added the comment you requested.

About the old/new back-end, I didn't even know there was such a thing, so I have no idea. Could you point me to the code and tests? I'll see if I can do it, but I'm pretty busy this week and next, so if you think it's faster/better for you to do it, just go ahead!

nouiz commented:
About the new back-end, this wiki page tell how to use it:

https://github.com/Theano/Theano/wiki/Converting-to-the-new-gpu-back-end%28gpuarray%29

nouiz commented:
I ported the change to the new back-end.

nouiz commented:
jenkins test this

lucasb-eyer commented:
Awesome, thank you for doing the port, I wasn't sure when I'd have time to do it!

nouiz commented:
I rebased the PR to make flake8 tests and amended the last commit to make this work.

nouiz commented:
jenkins test this

nouiz commented:
jenkins test this
